python/ray/data/tests/test_parquet.py

```diff
@@ -1176,6 +1176,26 @@ def test_write_num_rows_per_file(tmp_path, ray_start_regular_shared, num_rows_pe
         assert len(table) == num_rows_per_file
 
 
+@pytest.mark.parametrize(
+    "row_data",
+    [
+        [{"a": 1, "b": None}, {"a": 1, "b": 2}],
+        [{"a": None, "b": None}, {"a": 1, "b": 2}],
+        [{"a": 1, "b": 2}, {"a": 1, "b": "hi"}],
+    ],
+    ids=["row1_b_null", "row1_a_null", "row2_a_null"],
+)
+def test_write_auto_infer_nullable_fields(tmp_path, ray_start_regular_shared, row_data):
+    from ray.data._internal.logical.optimizers import _PHYSICAL_RULES
+    from ray.data._internal.logical.rules.operator_fusion import OperatorFusionRule
+
+    if OperatorFusionRule in _PHYSICAL_RULES:
+        _PHYSICAL_RULES.remove(OperatorFusionRule)
```

Comment:
Add a note why we're removing operator fusion here?

Reply:
Wait, why are we removing fusion?

We'd not need to do that

Reply:
When we don't do this - I think there will be only 1 block somehow. So the repro here didn't work, I guess we would need some other examples/repros if we don't remove the rule?

Reply:
> just add config to DataContext disabling it

@alexeykudinkin do you envision us adding a config for each optimization rule, or special-case operator fusion?

In any case, adding an interface for disabling optimization rules seems orthogonal to the goal of this PR, and can probably be handled as a follow-up?

Reply:
Tests should not be manipulating Data internals this way. Whether we split it into 2 PRs or do here, i'm not particularly concerned about.

Reply:
Yeah, I agree we should provide an abstraction for disabling optimization rule rather than directly modifying the list.

An alternative to adding a config to `DataContext` is to introduce a function `disable_logical_rule` to the optimizers module, analogous to `register_logical_rule`: https://github.com/ray-project/ray/blob/f6439e1c4b8808686ef98465e7372f77a7859fdd/python/ray/data/_internal/logical/optimizers.py#L33-L35

Reply:
Adding a draft PR here: https://github.com/ray-project/ray/pull/48617

Reply:
@rickyyx no need to block this PR on this, let's just reshape your test a bit:

 - Instead of using `ray.data.range` as source, create 2 parquet files -- 1 without nulls, another with nulls
 - Read both of these and then write out as single one

Reply:
@alexeykudinkin I might be missing something here, but I am not sure how I can force writing the 2 files with a single block w/o disabling the operator fusing. 

Something like below still only writes to the file with a single block (so there's technically no schema unification needed) 


```
    # Write each row to a separate file.
    for i, row in enumerate(row_data):
        ray.data.from_pandas(pd.DataFrame([row])).write_parquet(
            os.path.join(tmp_path, f"file_{i}.parquet")
        )

    # Read files and merge into a single file shouldn't error.
    ray.data.read_parquet(tmp_path).write_parquet(tmp_path, num_rows_per_file=2)
```

Reply:
> Instead of using ray.data.range as source, create 2 parquet files -- 1 without nulls, another with nulls
> Read both of these and then write out as single one

I don't think this'd reproduce the error. IIRC Ray Data will read both files in a single task, and then `BlockOutputBuffer` will combine the read data into a single block before passing it to the datasink

Reply:
Yeah, that might require some fidgeting to make it work.

Alternative path is to specify `num_cpus` which should make them diverge and hence avoid fusion.

Reply:
I was able to force with by changing the `target_max_block_size` if that's a better approach.